packages/models-library/src/models_library/utils/string_substitution.py

```diff
@@ -0,0 +1,108 @@
+""" Utils for substitutions in string templates
+
+"""
+import re
+from collections import UserDict
+from string import Template
+from typing import Any
+
+OSPARC_IDENTIFIER_PREFIX = "OSPARC_ENVIRONMENT"
+
+
+def upgrade_identifier(identifier: str) -> str:
+    """
+    Converts legacy or incompatible identifiers into official replacement identifier
+    """
+    identifier = re.sub(r"[{}%]", "", identifier)
+    identifier = re.sub(r"[.-]", "_", identifier)
+    identifier = identifier.upper()
+    if not identifier.startswith(OSPARC_IDENTIFIER_PREFIX):
+        identifier = f"{OSPARC_IDENTIFIER_PREFIX}_{identifier}"
+    return identifier
+
+
+# NOTE: includes % or %%
+_LEGACY_IDENTIFIER_RE_PATTERN = re.compile(r"%{1,2}([_a-z][_a-z0-9\.\-]*)%{1,2}")
+
+
+def substitute_all_legacy_identifiers(text: str) -> str:
+    """Substitutes all legacy identifiers found in the text by the new format expected in TemplateText
+
+    For instance:  '%%this-identifier%%' will be substituted by '$OSPARC_ENVIRONMENT_THIS_IDENTIFIER'
+    """
+
+    def _upgrade(match):
+        legacy_id = match.group(1)
+        legacy_id = upgrade_identifier(legacy_id)
+        return f"${legacy_id}"
+
+    return re.sub(_LEGACY_IDENTIFIER_RE_PATTERN, _upgrade, text)
+
+
+class TemplatePy311Mixin:
+    # NOTE: Remove these two in py3.11
+    # SEE https://github.com/python/cpython/blob/main/Lib/string.py#L144
+    #
```

Comment:
you can add a nice version check here that raises an error so you will not forget to remove this when upgrading to py11

Reply:
I will adapt it so it can get deactivated for python >=3.11